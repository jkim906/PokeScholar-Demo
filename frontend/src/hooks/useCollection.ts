import { useEffect } from 'react';
import { useUser } from '@clerk/clerk-react';
import { useCollectionStore } from '../stores/collectionStore';

/**
 * useCollection Hook
 * 
 * Manages the user's Pokemon card collection.
 * This hook provides access to collection data and methods
 * to load and filter the collection.
 * 
 * Features:
 * - Fetches and manages user's card collection
 * - Handles collection filtering
 * - Provides methods for loading collection data
 * - Automatically syncs with backend on mount
 * - Clears state on user sign out
 */
export const useCollection = () => {
  // Get user authentication state from Clerk
  const { user } = useUser();
  
  // Get collection store methods and state
  const {
    cards,
    allCards,
    filteredTotal,
    isLoading,
    error,
    loadCollection,
    setCards,
    setAllCards,
    setFilteredTotal,
    setError,
  } = useCollectionStore();

  // Load collection on mount and clear state on sign out
  useEffect(() => {
    if (user?.id) {
      loadCollection(user.id);
    } else {
      // Clear the state when user signs out
      setCards([]);
      setAllCards([]);
      setFilteredTotal(0);
      setError(null);
    }
  }, [user?.id, loadCollection, setCards, setAllCards, setFilteredTotal, setError]);

  return {
    cards,
    allCards,
    filteredTotal,
    isLoading,
    error,
    setCards,
    setAllCards,
    setFilteredTotal,
    // Wrap loadCollection to ensure user ID is available
    loadCollection: (filters?: { rarity?: string; type?: string }) => 
      user?.id ? loadCollection(user.id, filters) : Promise.reject('No user ID'),
  };
}; 