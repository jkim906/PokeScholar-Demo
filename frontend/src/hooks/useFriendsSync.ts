import { useEffect } from "react";
import { useUser } from "@clerk/clerk-react";
import { useFriendsStore } from "../stores/friendsStore";

/**
 * useFriendsSync is a custom hook that synchronizes the user's friends list
 * with the backend. It automatically fetches the friends list when the user is signed in
 * and updates the Zustand store with the latest data.
 *
 * This hook should be used in components that need to display or manage the user's friends.
 * It handles loading states and error handling automatically.
 */
export const useFriendsSync = () => {
  const { user } = useUser();
  const {
    friends,
    pendingRequests,
    isLoading,
    error,
    loadFriends,
    loadFriendsRequest,
    removeFriend,
    sendFriendRequest,
    acceptFriendRequest,
    declineFriendRequest,
    setFriends,
    setPendingRequests,
    setError,
  } = useFriendsStore();

  useEffect(() => {
    if (user?.id) {
      loadFriends(user.id);
      loadFriendsRequest(user.id);
    } else {
      // Clear the state when user signs out
      setFriends([]);
      setPendingRequests([]);
      setError(null);
      }
  }, [user?.id, loadFriends, loadFriendsRequest, setFriends, setPendingRequests, setError]);

  return {
    friends,
    pendingRequests,
    isLoading,
    error,
    loadFriends,
    loadFriendsRequest,
    removeFriend,
    sendFriendRequest,
    acceptFriendRequest,
    declineFriendRequest,
  };
};
