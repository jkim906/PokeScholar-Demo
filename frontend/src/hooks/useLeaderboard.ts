import { useEffect } from 'react';
import { useUser } from '@clerk/clerk-react';
import { useLeaderboardStore } from '../stores/leaderboardStore';

/**
 * useLeaderboard Hook
 * 
 * Manages leaderboard data and display preferences.
 * This hook provides access to leaderboard information and methods
 * to toggle between different leaderboard types.
 * 
 * Features:
 * - Fetches and manages leaderboard data
 * - Handles leaderboard type switching (session/coin based)
 * - Provides methods for updating leaderboard display
 * - Automatically syncs with backend on mount and type change
 */
export const useLeaderboard = () => {
  // Get user authentication state from Clerk
  const { user } = useUser();
  
  // Get leaderboard store methods and state
  const {
    leaderboardData,
    isLoading,
    error,
    leaderboardType,
    fetchLeaderboard,
    setLeaderboardType,
  } = useLeaderboardStore();

  // Fetch leaderboard data on mount and when type changes
  useEffect(() => {
    if (user?.id) {
      fetchLeaderboard(user.id);
    }
  }, [user?.id, fetchLeaderboard, leaderboardType]);

  // Toggle between session and point based leaderboards
  const toggleLeaderboardType = () => {
    setLeaderboardType(leaderboardType === 'session' ? 'coin' : 'session');
  };

  return {
    leaderboardData,
    isLoading,
    error,
    leaderboardType,
    toggleLeaderboardType,
    fetchLeaderboard,
  };
}; 