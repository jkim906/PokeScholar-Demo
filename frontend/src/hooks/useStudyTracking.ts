import { useEffect } from 'react';
import { useUser } from '@clerk/clerk-react';
import { useStudyTrackingStore } from '../stores/studyTrackingStore';

/**
 * useStudyTracking Hook
 * 
 * Manages study tracking data and statistics.
 * This hook provides access to study session information
 * and methods to track study progress.
 * 
 * Features:
 * - Fetches and manages weekly study statistics
 * - Tracks study session progress
 * - Provides methods for updating study data
 * - Automatically syncs with backend on mount
 */
export const useStudyTracking = () => {
  // Get user authentication state from Clerk
  const { user } = useUser();
  
  // Get study tracking store methods and state
  const {
    weeklyStats,
    isLoading,
    error,
    fetchWeeklyStats,
  } = useStudyTrackingStore();

  // Fetch study data on mount and when user ID changes
  useEffect(() => {
    if (user?.id) {
      fetchWeeklyStats(user.id);
    }
  }, [user?.id, fetchWeeklyStats]);

  return {
    weeklyStats,
    isLoading,
    error,
    fetchWeeklyStats: (userId: string) => fetchWeeklyStats(userId),
  };
}; 