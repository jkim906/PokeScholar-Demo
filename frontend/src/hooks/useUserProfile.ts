import { useEffect } from 'react';
import { useUser } from '@clerk/clerk-react';
import { useUserStore } from '../stores/userStore';

/**
 * useUserProfile Hook
 * 
 * Manages user profile data and card display preferences.
 * This hook provides access to user information and methods
 * to update profile settings and card display.
 * 
 * Features:
 * - Fetches and manages user profile information
 * - Handles card display preferences
 * - Provides methods for updating profile and card display
 * - Manages coin balance
 * - Automatically syncs with backend on mount
 */
export const useUserProfile = () => {
  // Get user authentication state from Clerk
  const { user } = useUser();
  
  // Get user store methods and state
  const {
    userInfo,
    cardDisplay,
    isLoading,
    error,
    fetchUserInfo,
    updateProfile,
    updateCardDisplay,
    fetchCardDisplay,
    addCoins,
    deductCoins,
  } = useUserStore();

  // Fetch user data on mount and when user ID changes
  useEffect(() => {
    if (user?.id) {
      fetchUserInfo(user.id);
      fetchCardDisplay(user.id);
    }
  }, [user?.id, fetchUserInfo, fetchCardDisplay]);

  return {
    userInfo,
    cardDisplay,
    isLoading,
    error,
    // Wrap methods to ensure user ID is available
    updateProfile: (profileData: { profileImage?: string }) => 
      user?.id ? updateProfile(user.id, profileData) : Promise.reject('No user ID'),
    updateCardDisplay: (cardIds: string[]) => 
      user?.id ? updateCardDisplay(user.id, cardIds) : Promise.reject('No user ID'),
    addCoins,
    deductCoins,
    fetchUserInfo: (userId: string) => fetchUserInfo(userId),
    fetchCardDisplay: (userId: string) => fetchCardDisplay(userId),
  };
}; 