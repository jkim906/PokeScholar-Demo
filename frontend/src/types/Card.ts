export interface Card {
  _id: string;
  name: string;
  rarity: string;
  types: string[];
  small: string;  // Direct image URL
  large: string;  // Direct image URL
  copies: number;
  collectedAt: string;
}