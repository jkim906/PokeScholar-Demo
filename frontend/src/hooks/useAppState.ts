import { useUser } from '@clerk/clerk-react';
import { useStudyTracking } from './useStudyTracking';
import { useLeaderboard } from './useLeaderboard';
import { useCollection } from '../hooks/useCollection';
import { useFriendsSync } from './useFriendsSync';
import { useUserProfile } from './useUserProfile';
import { useEffect, useCallback, useRef } from 'react';

/**
 * useAppState Hook
 * 
 * A central hook that manages and coordinates all application state.
 * This hook combines multiple stores and provides a unified interface
 * for accessing and updating application data.
 * 
 * Features:
 * - Manages user authentication state
 * - Coordinates data fetching across different stores
 * - Provides a single refresh mechanism for all data
 * - Handles initial data loading
 * - Manages leaderboard updates based on friends list changes
 */
export const useAppState = () => {
  // Get user authentication state from Clerk
  const { user, isSignedIn } = useUser();
  
  // Initialize all store hooks
  const studyTracking = useStudyTracking();
  const leaderboard = useLeaderboard();
  const collection = useCollection();
  const friends = useFriendsSync();
  const userProfile = useUserProfile();
  
  // Ref to prevent multiple simultaneous refreshes
  const isRefreshing = useRef(false);

  /**
   * refreshAllData
   * 
   * Refreshes all application data in parallel.
   * Uses a ref to prevent multiple simultaneous refreshes.
   * Only executes if user is signed in and not already refreshing.
   */
  const refreshAllData = useCallback(async () => {
    if (!isSignedIn || !user?.id || isRefreshing.current) return;
    
    try {
      isRefreshing.current = true;
      await Promise.all([
        userProfile.fetchUserInfo(user.id),
        userProfile.fetchCardDisplay(user.id),
        collection.loadCollection({}),
        studyTracking.fetchWeeklyStats(user.id),
        leaderboard.fetchLeaderboard(user.id),
        friends.loadFriends(user.id)
      ]);
    } catch (error) {
      console.error('Error refreshing data:', error);
    } finally {
      isRefreshing.current = false;
    }
  }, [isSignedIn, user?.id, userProfile, collection, studyTracking, leaderboard, friends]);

  // Initial data load when user signs in
  useEffect(() => {
    if (isSignedIn && user?.id) {
      refreshAllData();
    }
  }, [isSignedIn, user?.id]);

  // Refresh leaderboard when friends list changes
  useEffect(() => {
    if (userProfile.userInfo?._id) {
      leaderboard.fetchLeaderboard(userProfile.userInfo._id);
    }
  }, [friends.friends, userProfile.userInfo?._id]);

  return {
    user,
    studyTracking,
    leaderboard,
    collection,
    friends,
    userProfile,
    refreshAllData,
  };
}; 