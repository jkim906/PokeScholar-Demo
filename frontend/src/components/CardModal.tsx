import React from "react";
import {
  View,
  Image,
  Modal,
  Pressable,
  Text,
  ActivityIndicator,
  StyleSheet,
} from "react-native";
import { Card } from "../types/Card";
import { LinearGradient } from "expo-linear-gradient";

interface CardModalProps {
  selectedCard: Card | null;
  isImageLoading: boolean;
  onClose: () => void;
  onLoadStart: () => void;
  onLoad: () => void;
  testID?: string;
}

/**
 * CardModal Component
 *
 * Displays a full-screen modal with a detailed view of the selected card.
 * Features:
 * - Full-screen card image display
 * - Loading indicator while image loads
 * - Close button to dismiss modal
 * - Error handling for image loading
 *
 * @param {CardModalProps} props - Component props
 * @param {Card | null} props.selectedCard - The card to display in the modal
 * @param {boolean} props.isImageLoading - Loading state of the image
 * @param {Function} props.onClose - Callback when modal is closed
 * @param {Function} props.onLoadStart - Callback when image starts loading
 * @param {Function} props.onLoad - Callback when image finishes loading
 */
const CardModal: React.FC<CardModalProps> = ({
  selectedCard,
  isImageLoading,
  onClose,
  onLoadStart,
  onLoad,
  testID,
}) => {
  return (
    <Modal
      visible={selectedCard !== null}
      transparent
      animationType="fade"
      onRequestClose={onClose}
      testID={testID}
    >
      <Pressable style={styles.modalBackground} onPress={onClose}>
        {selectedCard?.large && (
          <View style={styles.cardContainer}>
            <Image
              source={{ uri: selectedCard.large }}
              style={styles.cardImage}
              onLoadStart={onLoadStart}
              onLoad={onLoad}
              onError={() => {
                console.error("Failed to load image:", selectedCard.large);
                onLoad();
              }}
              resizeMode="contain"
              testID="card-image"
            />
            <LinearGradient
              colors={[
                "rgba(255,255,255,0.4)",
                "rgba(255,255,255,0.2)",
                "rgba(255,255,255,0.1)",
                "rgba(255,255,255,0)",
              ]}
              locations={[0, 0.3, 0.6, 1]}
              start={{ x: 0, y: 0 }}
              end={{ x: 1, y: 1 }}
              style={styles.cardGradient}
            />
            {isImageLoading && (
              <View style={[styles.loadingContainer, styles.cardImage]}>
                <ActivityIndicator
                  testID="loading-indicator"
                  size="large"
                  color="#007BFF"
                />
              </View>
            )}
          </View>
        )}
        <Text style={styles.tapHint}>Tap anywhere to close</Text>
      </Pressable>
    </Modal>
  );
};

const styles = StyleSheet.create({
  modalBackground: {
    flex: 1,
    backgroundColor: "rgba(0,0,0,0.8)",
    justifyContent: "center",
    alignItems: "center",
  },
  cardContainer: {
    width: 300,
    height: 420,
    alignItems: "center",
    justifyContent: "center",
    position: "relative",
    borderRadius: 12,
    overflow: "hidden",
  },
  cardImage: {
    width: "100%",
    height: "100%",
    borderRadius: 12,
  },
  cardGradient: {
    position: "absolute",
    top: 0,
    left: 0,
    right: 0,
    bottom: 0,
    opacity: 1,
    borderRadius: 12,
  },
  tapHint: {
    position: "absolute",
    bottom: 20,
    color: "#fff",
    fontSize: 14,
  },
  loadingContainer: {
    position: "absolute",
    justifyContent: "center",
    alignItems: "center",
    backgroundColor: "rgba(240,240,240,0.7)",
    borderRadius: 12,
  },
});

export default CardModal;
