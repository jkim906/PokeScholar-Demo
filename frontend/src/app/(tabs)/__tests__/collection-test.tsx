import React from 'react';
import { render, fireEvent, waitFor, act } from '@testing-library/react-native';
import Collection from '../collection';
import { useUser } from '@clerk/clerk-react';
import { useAppState } from '../../../hooks/useAppState';
import { Card } from '../../../types/Card';
import CardModal from '@/src/components/CardModal';

// Mock the hooks
jest.mock('@clerk/clerk-react');
jest.mock('../../../hooks/useAppState');

// Mock the font loading
jest.mock('expo-font', () => ({
    loadAsync: jest.fn(),
    isLoaded: jest.fn().mockReturnValue(true),
    isLoading: jest.fn().mockReturnValue(false),
  }));
  

// Mock the navigation hook
jest.mock('@react-navigation/native', () => ({
  useFocusEffect: jest.fn((callback) => callback()),
}));

// Mock the components
jest.mock('../../../components/SortModal', () => {
  const React = require('react');
  const { View, Text, TouchableOpacity } = require('react-native');
  
  return function MockSortModal({ 
    visible, 
    onClose, 
    sortBy, 
    sortDirection, 
    onSortChange 
  }: {
    visible: boolean;
    onClose: () => void;
    sortBy: string | null;
    sortDirection: 'asc' | 'desc';
    onSortChange: (option: string) => void;
  }) {
    if (!visible) return null;
    return React.createElement(View, { testID: "sort-modal" }, [
      React.createElement(TouchableOpacity, { 
        key: 'recent',
        onPress: () => onSortChange('Recent')
      }, React.createElement(Text, null, 'Recent')),
      React.createElement(TouchableOpacity, { 
        key: 'type',
        onPress: () => onSortChange('Type')
      }, React.createElement(Text, null, 'Type')),
      React.createElement(TouchableOpacity, { 
        key: 'rarity',
        onPress: () => onSortChange('Rarity')
      }, React.createElement(Text, null, 'Rarity')),
      React.createElement(TouchableOpacity, { 
        key: 'duplicates',
        onPress: () => onSortChange('Duplicates')
      }, React.createElement(Text, null, 'Duplicates'))
    ]);
  };
});

jest.mock('../../../components/CardGridItem', () => {
  const React = require('react');
  const { View, Text, TouchableOpacity, Image } = require('react-native');
  
  return function MockCardGridItem({ 
    card, 
    onPress 
  }: {
    card: Card;
    onPress: (card: Card) => void;
  }) {
    return React.createElement(TouchableOpacity, { 
      testID: `card-${card._id}`,
      onPress: () => onPress(card)
    }, [
      React.createElement(Image, { 
        key: 'image',
        source: { uri: card.small }
      }),
      React.createElement(Text, { key: 'name' }, card.name),
      card.copies > 1 && React.createElement(Text, { key: 'copies' }, `x${card.copies}`)
    ].filter(Boolean));
  };
});

// Test the Collection component
describe('Collection Component', () => {
  const mockCards = [
    {
      _id: '1',
      name: 'Pikachu',
      rarity: 'Common',
      types: ['Electric'],
      copies: 2,
      small: 'pikachu-small.jpg',
      large: 'pikachu-large.jpg',
      collectedAt: '2024-03-10T00:00:00.000Z',
    },
    {
      _id: '2',
      name: 'Charizard',
      rarity: 'Rare',
      types: ['Fire'],
      copies: 1,
      small: 'charizard-small.jpg',
      large: 'charizard-large.jpg',
      collectedAt: '2024-03-09T00:00:00.000Z',
    },
    {
      _id: '3',
      name: 'Bulbasaur',
      rarity: 'Common',
      types: ['Grass'],
      copies: 3,
      small: 'bulbasaur-small.jpg',
      large: 'bulbasaur-large.jpg',
      collectedAt: '2024-03-11T00:00:00.000Z',
    },
    {
      _id: '4',
      name: 'Squirtle',
      rarity: 'Common',
      types: ['Electric'],
      copies: 1,
      small: 'squirtle-small.jpg',
      large: 'squirtle-large.jpg',
      collectedAt: '2024-03-12T00:00:00.000Z',
    }
  ];

  // Mock the loadCollection function
  const mockLoadCollection = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
    // Mock useUser hook
    (useUser as jest.Mock).mockReturnValue({
      user: { id: '123' },
      isSignedIn: true,
    });

    // Mock useAppState hook
    (useAppState as jest.Mock).mockReturnValue({
      collection: {
        cards: mockCards,
        filteredTotal: 3,
        isLoading: false,
        error: null,
        loadCollection: mockLoadCollection,
      },
    });
  });

  
  it('renders correctly when signed in', () => {
    const { getByText, getByPlaceholderText, getByTestId } = render(<Collection />);
    
    // Check if basic elements are rendered
    expect(getByPlaceholderText('Search cards...')).toBeTruthy();
    expect(getByText('All')).toBeTruthy();
    expect(getByText('Common')).toBeTruthy();
    expect(getByText('Rare')).toBeTruthy();
    expect(getByText('Sort: Recent')).toBeTruthy();

    // Check if cards are rendered
    expect(getByTestId('card-1')).toBeTruthy();
    expect(getByTestId('card-2')).toBeTruthy();
    expect(getByTestId('card-3')).toBeTruthy();
  });

  it('shows sign in message when not signed in', () => {
    (useUser as jest.Mock).mockReturnValue({
      user: null,
      isSignedIn: false,
    });

    const { getByText } = render(<Collection />);
    expect(getByText('Please sign in to view your collection')).toBeTruthy();
  });

  it('loads collection when screen is focused', () => {
    render(<Collection />);
    expect(mockLoadCollection).toHaveBeenCalledWith({});
  });

  it('loads collection with rarity filter when rarity is selected', async () => {
    const { getByText } = render(<Collection />);
    
    // Click on Rare filter
    fireEvent.press(getByText('Rare'));
    
    expect(mockLoadCollection).toHaveBeenCalledWith({ rarity: 'Rare' });
  });

  it('filters cards by search text', async () => {
    const { getByPlaceholderText, getByTestId, queryByTestId } = render(<Collection />);
    const searchInput = getByPlaceholderText('Search cards...');

    // Type in search
    fireEvent.changeText(searchInput, 'Pikachu');
    
    // Wait for filtering to complete
    await waitFor(() => {
      expect(getByTestId('card-1')).toBeTruthy();
      expect(queryByTestId('card-2')).toBeNull();
      expect(queryByTestId('card-3')).toBeNull();
    });
  });

  it('shows loading state', () => {
    (useAppState as jest.Mock).mockReturnValue({
      collection: {
        cards: [],
        filteredTotal: 0,
        isLoading: true,
        error: null,
        loadCollection: mockLoadCollection,
      },
    });

    const { getByTestId } = render(<Collection />);
    expect(getByTestId('loading-indicator')).toBeTruthy();
  });

  it('shows error state', () => {
    (useAppState as jest.Mock).mockReturnValue({
      collection: {
        cards: [],
        filteredTotal: 0,
        isLoading: false,
        error: 'Failed to load collection',
        loadCollection: mockLoadCollection,
      },
    });

    const { getByText } = render(<Collection />);
    expect(getByText('Failed to load collection')).toBeTruthy();
  });

  it('opens sort modal and changes sort option', async () => {
    const { getByText, getByTestId } = render(<Collection />);
    
    // Open sort modal
    fireEvent.press(getByText('Sort: Recent'));
    
    // Check if modal is rendered
    expect(getByTestId('sort-modal')).toBeTruthy();
    
    // Select Rarity sort
    fireEvent.press(getByText('Rarity'));
    
    // Verify the cards are sorted by rarity
    const sortedCards = [...mockCards].sort((a, b) => 
      ['Common', 'Uncommon', 'Rare', 'Illustration Rare', 'Double Rare', 'Special Illustration Rare']
        .indexOf(b.rarity) - ['Common', 'Uncommon', 'Rare', 'Illustration Rare', 'Double Rare', 'Special Illustration Rare']
        .indexOf(a.rarity)
    );
    expect(sortedCards[0].name).toBe('Charizard');
  });

  it('shows duplicate count on cards', () => {
    const { getByTestId } = render(<Collection />);
    
    // Check if duplicate count is shown for cards with copies > 1
    const pikachuCard = getByTestId('card-1');
    const bulbasaurCard = getByTestId('card-3');
    
    expect(pikachuCard).toHaveTextContent(/x2/);

    expect(bulbasaurCard).toHaveTextContent(/x3/);
  });

  it('shows empty state message when no cards are in the collection', () => {
    (useAppState as jest.Mock).mockReturnValue({
      collection: {
        cards: [],
        filteredTotal: 0,
        isLoading: false,
        error: null,
        loadCollection: mockLoadCollection,
      },
    });
  
    const { getByText } = render(<Collection />);
    expect(getByText('No cards found.')).toBeTruthy();
  });
  
  it('sorts cards by recent collection date', async () => {
    const { getByText, getByTestId } = render(<Collection />);
    
    // Open sort modal
    fireEvent.press(getByText('Sort: Recent'));
    
    // Click on 'Recent' sort option
    fireEvent.press(getByText('Recent'));
    
    // Verify that cards are sorted by collection date (most recent first)
    const sortedCards = [...mockCards].sort((a, b) => new Date(b.collectedAt).getTime() - new Date(a.collectedAt).getTime());
    expect(sortedCards[0].name).toBe('Squirtle');
  });
  
  it('sorts by type and then by name when types are equal', () => {
    const sorted = [...mockCards].sort((a, b) => {
      let comparison = (a.types[0] || '').localeCompare(b.types[0] || '');
      if (comparison === 0) comparison = a.name.localeCompare(b.name);
      return comparison;
    });
  
    expect(sorted).toMatchSnapshot(); 
  });

  it('sorts cards by type (lexical order, fallback to name)', async () => {
    const { getByText, getAllByTestId } = render(<Collection />);
  
    // Open sort modal
    fireEvent.press(getByText('Sort: Recent'));
  
    // Select Type
    fireEvent.press(getByText('Type'));
    await waitFor(() => {
      const cardNames = getAllByTestId(/^card-/).map(card => {
        const nameElement = card.children[1];
        if (typeof nameElement === 'string') return nameElement;
        return nameElement.props.children;
      });
      const expectedOrder = ['Pikachu', 'Squirtle', 'Charizard', 'Bulbasaur']; 
      expect(cardNames).toEqual(expectedOrder);
    });
  });
  
  it('sorts cards by number of duplicates (copies descending)', async () => {
    const { getByText, getAllByTestId } = render(<Collection />);
  
    // Open sort modal
    fireEvent.press(getByText('Sort: Recent'));
  
    // Select Duplicates
    fireEvent.press(getByText('Duplicates'));
    await waitFor(() => {
      const cardNames = getAllByTestId(/^card-/).map(card => {
        const nameElement = card.children[1];
        if (typeof nameElement === 'string') return nameElement;
        return nameElement.props.children;
      });
      const expectedOrder = ['Bulbasaur', 'Pikachu', 'Charizard', 'Squirtle']; // x3 > x2 > x1 > x1
      expect(cardNames).toEqual(expectedOrder);
    });
  });
  

  it('sorts cards by number of copies (descending)', () => {
    const sorted = [...mockCards].sort((a, b) => b.copies - a.copies);
    expect(sorted.map(c => c.copies)).toEqual([3, 2, 1, 1]);
  });

  it('opens and closes the CardModal', async () => {
    const { getByTestId, queryByTestId } = render(<Collection />);
    
    // Modal should not be shown initially
    expect(queryByTestId('card-image')).toBeNull();
  
    // Press a card to open modal
    fireEvent.press(getByTestId('card-1'));
    expect(getByTestId('card-image')).toBeTruthy();
  
    // Press the modal overlay to close
    fireEvent.press(getByTestId('card-image'));
  
    // Modal should disappear after state update
    await waitFor(() => {
      expect(queryByTestId('card-image')).toBeNull();
    }, { timeout: 1000 });
  });
  

  it('handles image load states', async () => {
    const { getByTestId, queryByTestId } = render(<Collection />);
    
    // Open modal
    fireEvent.press(getByTestId('card-1'));
  
    const cardImage = getByTestId('card-image');
  
    // Simulate image loading start
    fireEvent(cardImage, 'loadStart');
    expect(getByTestId('loading-indicator')).toBeTruthy();
  
    // Simulate image loading complete
    fireEvent(cardImage, 'load');
  
    await waitFor(() => {
      expect(queryByTestId('loading-indicator')).toBeNull();
    }, { timeout: 1000 });
  });
  
});
