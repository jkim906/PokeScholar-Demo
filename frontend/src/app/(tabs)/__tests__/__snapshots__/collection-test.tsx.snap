// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collection Component sorts by type and then by name when types are equal 1`] = `
[
  {
    "_id": "1",
    "collectedAt": "2024-03-10T00:00:00.000Z",
    "copies": 2,
    "large": "pikachu-large.jpg",
    "name": "Pikachu",
    "rarity": "Common",
    "small": "pikachu-small.jpg",
    "types": [
      "Electric",
    ],
  },
  {
    "_id": "4",
    "collectedAt": "2024-03-12T00:00:00.000Z",
    "copies": 1,
    "large": "squirtle-large.jpg",
    "name": "Squirtle",
    "rarity": "Common",
    "small": "squirtle-small.jpg",
    "types": [
      "Electric",
    ],
  },
  {
    "_id": "2",
    "collectedAt": "2024-03-09T00:00:00.000Z",
    "copies": 1,
    "large": "charizard-large.jpg",
    "name": "Charizard",
    "rarity": "Rare",
    "small": "charizard-small.jpg",
    "types": [
      "Fire",
    ],
  },
  {
    "_id": "3",
    "collectedAt": "2024-03-11T00:00:00.000Z",
    "copies": 3,
    "large": "bulbasaur-large.jpg",
    "name": "Bulbasaur",
    "rarity": "Common",
    "small": "bulbasaur-small.jpg",
    "types": [
      "Grass",
    ],
  },
]
`;
